### PR TITLE
Check int for input and None for grad

### LINKED ISSUE #999: 1.7.0 debug mode bug
Hi,

I tested the debug model on one of my existing models and got:

```
Traceback (most recent call last):
  File "./train.py", line 53, in <module>
    loss.backward()
  File "/home/eddie/envs/deep/local/lib/python2.7/site-packages/chainer/variable.py", line 355, in backward
    for gx in gxs):
  File "/home/eddie/envs/deep/local/lib/python2.7/site-packages/chainer/variable.py", line 355, in <genexpr>
    for gx in gxs):
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

Here gx is 
```
[[ 0.0001472   0.00014604  0.00015193 ...,  0.00015092  0.0001509
   0.00014626]
 [ 0.00014647  0.0001442   0.00015269 ...,  0.00015215  0.00015142
   0.00014504]
 [ 0.00014855  0.00014735  0.00015144 ...,  0.00015121  0.00015061
   0.00014711]
 ...,
 [ 0.00014409  0.00014379  0.00015338 ...,  0.00015306  0.00015099
   0.00014355]
 [ 0.00014459  0.00014202  0.00015292 ...,  0.00015324  0.00015215
   0.00014102]
 [ 0.00014621  0.00014452  0.00015207 ...,  0.00015231  0.00015252
   0.00014397]]
```

composed of np.float32

### PR DESCRIPTION
**Pull request by unnonouno**
- only check float inputs
- ignore none for gradients

fix #999

### CODE REVIEW

File: tests/chainer_tests/test_function.py [hunk 1]
```diff
@@ -343,11 +348,25 @@ def test_debug_forward_gpu(self):
             return_value=(cuda.to_gpu(self.nan),))
         self.check_debug_forward(cuda.to_gpu(self.one))
 
-    def check_debug_backward(self, x_data):
+    def test_debug_forward_int_cpu(self):
+        self.f.forward_cpu = mock.MagicMock(return_value=(self.int_one,))
+        self.check_debug_forward(self.one, valid=True)
```
Comment (delta2323) [inline tests/chainer_tests/test_function.py]: Can we make `test_debug_forward_cpu` and `test_debug_forward_int_cpu` parameterized with respect to `return_value` and `valid`?

File: tests/chainer_tests/test_function.py [hunk 2]
```diff
@@ -362,5 +381,18 @@ def test_debug_backward_gpu(self):
             return_value=(cuda.to_gpu(self.nan),))
         self.check_debug_backward(cuda.to_gpu(self.one))
 
+    def test_debug_backward_none_cpu(self):
+        self.f.forward_cpu = mock.MagicMock(return_value=(self.one,))
+        self.f.backward_cpu = mock.MagicMock(return_value=(None,))
+        self.check_debug_backward(self.one, valid=True)
+
```
Comment (delta2323) [inline tests/chainer_tests/test_function.py]: Can we make make `test_debug_backward_cpu` and `test_debug_backward_none_cpu` parameterized with respect to `return_value` of backward and `valid`? (We may need to separate forward tests (e.g. `test_debug_forward_cpu`) and backward tests (e.g. `test_debug_backward_cpu`)).

File: chainer/function.py [hunk 3]
```diff
@@ -121,7 +121,8 @@ def __call__(self, *inputs):
             hook.forward_postprocess(self, in_data)
 
         if chainer.is_debug():
-            if any(cuda.get_array_module(out).isnan(out).any()
+            if any(out.dtype.kind == 'f' and
```
Comment (delta2323) [inline chainer/function.py]: Why do we check only float values?

Comment (unnonouno) [inline chainer/function.py]: int values cannot be nan.

**Comment by delta2323** [comment]
LGTM :+1:

**Pull request merged**